### DISCUSSION TITLE
Cancel workflow instance run multiple times

### DISCUSSION THREAD

**Discussion by SpartanNico94** [elsa-workflows/elsa-core#5498] [Q&A] [answered]
Hi, I'm trying to delete/cancel the execution of a workflow if a condition is met. 
I found the following code and now I can stop the workflow and delete the instance from database, but Elsa (2.12.0) runs the same workflow 5 times before stopping. 

NOTE: this code is in my WorkflowExecuting event handler
``` c sharp
if(condition){
    instance.WorkflowStatus = WorkflowStatus.Cancelled; 
    await _workflowInstanceStore.SaveAsync(instance, cancellationToken);
    await _workflowInstanceCanceller.CancelAsync(workflowInstanceId, cancellationToken);
    //assignment only for debug...
    var result = await _workflowInstanceDeleter.DeleteAsync(workflowInstanceId, cancellationToken);
}
```

I'm using AzureServiceBus to handle messages and it looks that the message that triggers the workflow is not "consumed" after the first "cancel" operation, but after 5 times it goes on dead-letter. 
Is there something I'm missing with the configuration between Elsa and Azure? 
Thanks for the help!

**Comment by glime-ai** [comment]
To address the issue you're experiencing with the workflow being executed multiple times before stopping, it's important to understand how message handling works with Azure Service Bus and Elsa. When a message is picked up by a consumer (in this case, Elsa), it's expected that the consumer completes the message if it has been successfully processed, or abandons it if not. If neither of these actions is taken, the message will be retried until the maximum delivery count is reached, after which it will be moved to the dead-letter queue.

In your case, it seems that the message triggering the workflow is not being completed or abandoned correctly after you cancel the workflow instance. This could be why the message is retried multiple times.

Here are a few things you can check or try:

1. **Ensure Message Completion**: Make sure that after cancelling the workflow, the message is also completed so that it's not retried. This might involve interacting with the Azure Service Bus API directly to complete the message.

2. **Check the Delivery Count**: Review the configuration for the maximum delivery count on the Azure Service Service Bus subscription to understand after how many attempts a message will be dead-lettered.

3. **Dead-Letter Handling**: If the message goes to the dead-letter queue, you may need to handle it separately, either by resubmitting it after fixing any issues or by discarding it if it's no longer needed.

4. **PeekLock Duration**: Investigate the PeekLock duration setting for your Azure Service Bus. If the lock duration is too short and the workflow takes longer to cancel, the message may be retried.

5. **Elsa and Azure Service Bus Configuration**: Review the configuration between Elsa and Azure Service Bus to ensure that it aligns with your workflow requirements and message handling expectations.

6. **Elsa Version**: You mentioned using Elsa 2.12.0. It might be worth checking if there are any updates or bug fixes related to Azure Service Bus in newer versions of Elsa that could resolve your issue.

7. **Elsa Documentation and Issues**: Review the Elsa documentation and search for any similar issues reported by other users. Sometimes, the community or the maintainers may have provided solutions or workarounds for similar problems.

8. **Logging and Debugging**: Add more detailed logging around the message handling and workflow cancellation process to get a better understanding of what's happening under the hood.

If you continue to face issues, consider reaching out to the Elsa community or filing an issue on the Elsa GitHub repository with detailed information about your problem for further assistance.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Comment by SpartanNico94** [comment, marked as answer]
After looking at the source code, setting the status as Cancelled during the "WorkflowExecuting" handler makes Elsa to throw an exception. I had to place my logic to stop the workflow in my ValidateWorkflowExecution Handler.
